### DISCUSSION TITLE
Using egui with async functions

### DISCUSSION THREAD

**Discussion by melody-rs** [emilk/egui#2010] [Q&A] [answered]
## EDIT
The easiest way to use async code is to use the `poll_promise` crate - you can spawn a task with the same code on wasm/native and poll for its completion in the Ui code.

---

I've been working on an egui app for a few days now and I've run into an annoying hitch, So far, I've been targeting both native and wasm32 using eframe, with the goal of using the new JavaScript [File System Access API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API) to handle IO. Unfortunately for me, all the functions in it are async, which poses a problem for me since all my code is synchronous. 

I've also noticed that `WASI` has functions for filesystem, which would be amazing to use, but I don't know how I could get that to work with trunk or the eframe tamplate.

I'd ideally like to keep it synchronous, but sadly it seems that isn't possible. I saw [this thread](https://github.com/emilk/egui/discussions/521) which suggests using [Pollster](https://github.com/zesterer/pollster) but that ends up panicking with `condvar wait not supported`.

What should I do?

**Comment by notgull** [comment]
If you enable `target_feature = "atomics"`, that should ensure condvars work, at least on web environments.

**Reply by melody-rs** [reply]
I tried that actually, and it didn't work. It panics with another error `Atomics.wait cannot be called in this context` which is presumably from blocking the main thread.

**Comment by emilk** [comment, marked as answer]
Have you tried [`wasm_bindgen_futures`](https://docs.rs/wasm-bindgen-futures/latest/wasm_bindgen_futures/) ?

**Reply by melody-rs** [reply]
I have tried that out and it works, but sadly I need to work with futures that return values and wasm_bindgen_futures doesn't seem to support that from what I can tell

**Reply by ar37-rs** [reply]
Have you tried using [download_image](https://github.com/emilk/egui/blob/master/examples/download_image/src/main.rs) example technique which is using `poll_promise` to extract the value out of async process? and the try combining with `wasm_bindgen_futures`

**Reply by melody-rs** [reply]
> Have you tried using [download_image](https://github.com/emilk/egui/blob/master/examples/download_image/src/main.rs) example technique which is using `poll_promise` to extract the value out of async process? and the try combining with `wasm_bindgen_futures`

I am running in to a lot of problems doing this. I had to refactor a *lot* of my code to get to this point. I probably made some bad decisions which has lead me to this point, apologies. 

I have a struct `Filesystem` with an async function I need to call.
```rs
pub struct Filesystem {
    project_path: Option<PathBuf>,
    data_cache: Option<super::data_cache::DataCache>,
}
impl Filesystem {
// ...
    pub async fn try_open_project(&mut self) {
        if let Some(mut path) = rfd::FileDialog::default()
            .add_filter("project file", &["rxproj", "lum"])
            .pick_file()
        {
            path.pop(); // Pop off filename
            self.load_project(path)
        }
    }
// ...
}
```
This `Filesystem` struct is stored as a member of my `App` struct from eframe.
```rs
pub struct App {
    filesystem: Rc<RefCell<Filesystem>>,
    windows: Rc<RefCell<Windows>>,
    top_bar: crate::gui::top_bar::TopBar
}
```
In my update function I end up passing `Filesystem` off to windows/widgets to get updated.
```rs
// ...
let mut update_info = UpdateInfo {
    filesystem: self.filesystem.clone(),
    windows: self.windows.clone()
};

// We want the top menubar to be horizontal. Without this it would fill up vertically.
ui.horizontal_wrapped(|ui| {
    // Turn off button frame.
    ui.visuals_mut().button_frame = false;
    // Show the bar
    self.top_bar.ui(&mut update_info, ui, frame);
});

self.windows.borrow_mut().update(ctx, &mut update_info);
```
This works fine, but the problems start when trying to call `Filesystem::try_open_project()`.
(This is in a `TopBar` struct which stores a promise like the example you linked)
```rs
if ui.button("Open Project").clicked() {
   let promise = Promise::spawn_thread("file picker", move || async {
       filesystem.try_open_project().await;
   });
   self.project_open_promise = Some(promise);
}
```
Rust complains that the type of `TopBar::project_open_promise` does not match what I'm assigning:
```
error[E0308]: mismatched types
   --> src/gui/top_bar.rs:34:54
    |
31  |                       let promise = Promise::spawn_thread("file picker", move || async {
    |  ______________________________________________________________________________________-
32  | |                         filesystem.try_open_project().await;
33  | |                     });
    | |_____________________- the found `async` block
34  |                       self.project_open_promise = Some(promise);
    |                                                   ---- ^^^^^^^ expected `()`, found opaque type
    |                                                   |
    |                                                   arguments to this enum variant are incorrect
    |
   ::: /home/lily/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:72:43
    |
72  |   pub const fn from_generator<T>(gen: T) -> impl Future<Output = T::Return>
    |                                             ------------------------------- the found opaque type
    |
    = note: expected struct `Promise<()>`
               found struct `Promise<impl Future<Output = ()>>`
```
I tried changing the types to match but rust complains about using a trait as a generic argument. I can't simply add `dyn` here either because that requires a `Box`, which completely changes the type...

I tried wrapping `Filesystem` in `Arc<Mutex<T>>` instead, although that didn't work too (``` the trait `std::marker::Send` is not implemented for `std::sync::MutexGuard<'_, Filesystem>```). At that point I decided I was probably doing something wrong and should ask for more help.

**Reply by ar37-rs** [reply]
@Speak2Erase try using block_on, whether using future, pollster or whatever the executor, like so:  

```rust
if ui.button("Open Project").clicked() {
   let promise = Promise::spawn_thread("file picker", move || {
       pollster::block_on(async {
           filesystem.try_open_project().await;
       });
    });
   self.project_open_promise = Some(promise);
}
```

**Reply by melody-rs** [reply]
That works fine for native but not on webassembly. :pensive: 
![image](https://user-images.githubusercontent.com/54482069/188730163-6fd210b0-6064-4f40-9b1b-8524f01cffb6.png)

**Reply by ar37-rs** [reply]
Don't use spawn thread for wasm, but the technique as mentioned above.

**Reply by melody-rs** [reply]
Still doesn't work, but with a compiler error instead.
```
[E0277]: `[closure@src/gui/top_bar.rs:30:56: 35:22]` is not a future
   --> src/gui/top_bar.rs:30:35
    |
30  |                     let promise = Promise::spawn_async(move || {
    |                                   ^^^^^^^^^^^^^^^^^^^^ `[closure@src/gui/top_bar.rs:30:56: 35:22]` is not a future
    |
    = help: the trait `Future` is not implemented for `[closure@src/gui/top_bar.rs:30:56: 35:22]`
    = note: [closure@src/gui/top_bar.rs:30:56: 35:22] must be a future or must implement `IntoFuture` to be awaited
note: required by a bound in `Promise::<T>::spawn_async`
   --> /home/lily/.cargo/registry/src/github.com-1ecc6299db9ec823/poll-promise-0.1.0/src/promise.rs:155:37
    |
155 |     pub fn spawn_async(future: impl std::future::Future<Output = T> + 'static + Send) -> Self {
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Promise::<T>::spawn_async`
```

**Reply by melody-rs** [reply]
@ar37-rs 
> @Speak2Erase try using block_on, whether using future, pollster or whatever the executor, like so:
> 
> ```rust
> if ui.button("Open Project").clicked() {
>    let promise = Promise::spawn_thread("file picker", move || {
>        pollster::block_on(async {
>            filesystem.try_open_project().await;
>        });
>     });
>    self.project_open_promise = Some(promise);
> }
> ```

I had to rewrite my code quite a lot to get this to compile, and...
![image](https://user-images.githubusercontent.com/54482069/189030136-fe545920-8b2e-4bdb-8293-467f6285502f.png)

**Reply by emilk** [reply]
Use `Promise::spawn_async` with the `web` feature as covered by the docs: https://docs.rs/poll-promise/latest/poll_promise/#web

**Reply by melody-rs** [reply]
Bit late, but that's what I've been doing!

**Reply by jonboh** [reply]
@emilk did you mean `Promise::spawn_local`? `Promise::spawn_async` is only available on `async-std`, `tokio` and `smol`

**Reply by Lubba-64** [reply]
Ran into this in 2024. ouch.

**Reply by Lubba-64** [reply]
Nevermind, for anyone stumbling across this thread, the best I could do to extract the value out of the spawn_local context was to use this pattern:
```rust
#[cfg(target_arch = "wasm32")]
pub struct Task<T>(Rc<Cell<Option<thread::Result<T>>>>);

#[cfg(target_arch = "wasm32")]
impl<T: 'static> Task<T> {
    pub fn spawn<F: 'static + Future<Output = T>>(future: F) -> Self {
        let sender = Rc::new(Cell::new(None));
        let receiver = sender.clone();
        wasm_bindgen_futures::spawn_local(async move {
            let future = panic::AssertUnwindSafe(future).catch_unwind();
            sender.set(Some(future.await));
        });
        Self(receiver)
    }
    pub fn take_output(&self) -> Option<thread::Result<T>> {
        self.0.take()
    }
}

```

This sucks but it works so I am moving on with my life.

**Reply by dream-dasher** [reply]
Oof.  Thanks @Lubba-64 
One of the big sells of egui is being able to share apps on the web -- being able to use rust knowledge and still make something you can share with non-technical people.  But the wasm_bindgen_futures is basically a way to translate to javascript promises which seems to rather demand a new skillset to use with confidence, in addition to splitting the app.

Would be great to just use `#[tokio::main(flavor = "current_thread")]` and `tokio::spawn::task()`, with a `tokio::yield_now().await` or even `tokio::time::sleep(Duration::from_millis(10).await` at the bottom of the event loop.

Naive play with this does not work (of course): `eframe::run_native()` pops up a synchronous loop so we can spawn tasks, but the loop blocks them from ever operating in the thread.

(I thought perhaps there'd be a way to run sync code that runs asynchronous code that swapped tasks during the await, but it seems the relevant pieces of the Tokio runtime never get to run.  Which ... probably makes sense, though I need to familiarize myself better with both.)

-- **Has anyone tried** just ... doing a dirty fork of egui and running the event loop as asynch?
Would be grand to have a **native** and **wasm** version of the same code that just switch from `"multi_thread"` to `"current_thread"` -- would make testing, maintenance, development, and brain overhead all much easier.

(And, at least for me, there are quite a few utilities that would be useful that benefit from making enough http calls that async would be useful -- but I'd love to *not* swap my entire async approach just for the wasm version of the app.)

**Reply by melody-rs** [reply]
The easiest way is to use the `poll_promise` crate - you can spawn a task with the same code on wasm/native and poll for its completion in the Ui code

**Comment by ar37-rs** [comment]
👍
